IPIP-0383: Compact Denylist Format (#383)
ipfs/specs: COMPACT_DENYLIST_FORMAT.md

Same as with the `/ipfs/` rule, blocks IPNS paths starting with the given path prefix. `/*` is equivalent to the empty string, so `/abc/*` == `/abc*`.

Blocking layer recommendation: Resolver.

##### `/\<path\>` `/\<path\>/*` `/\<path\>*`

Block solely by looking at the path component, and ignoring the CID/IPNS parts, before resolving.

This blocks all the paths matching exactly or having the same prefix as the one in the rule:

- `/my/path`: blocks any item that tries to resolve `/my/path`, regardless of the CID used.
- `/my/path*` and `/my/path/*`: blocks any paths that contain the prefix `/my/path`.

Blocking layer recommendation: Resolver.

##### `//\<double_hash\>`

Blocks a double-hashed item, which can be:

- The hash of a CIDv1base32[+path]: legacy badbits, block-by-cid format
- The hash of an IPNS path `/ipns/*`.
- The hash of a CIDv0[+path]

Blocking layer recommendation: Resolver + BlockService.

In order to check for a matching rule, the Resolver should:

- IPFS path: convert the CID to v1base32 and hash the path without the `/ipfs/` prefix.
- IPFS path: convert the CID to v0 and hash the path without the `/ipfs/` prefix.

> - IPNS path: hash the path "as is".

Author: yes, this is correct.

```
version: 1
name: IPFSCorp blocking list
description: A collection of bad things we have found in the universe
author: abuse-ipfscorp@example.com
hints:
  hint: value
  hint2: value2
---
# Blocking by CID - blocks wrapped multihash.
# Does not block subpaths.
/ipfs/bafybeihvvulpp4evxj7x7armbqcyg6uezzuig6jp3lktpbovlqfkuqeuoq

# Block all subpaths
/ipfs/QmdWFA9FL52hx3j9EJZPQP1ZUH8Ygi5tLCX2cRDs6knSf8/*

# Block some subpaths (equivalent rules)
/ipfs/Qmah2YDTfrox4watLCr3YgKyBwvjq8FJZEFdWY6WtJ3Xt2/test*
/ipfs/QmTuvSQbEDR3sarFAN9kAeXBpiBCyYYNxdxciazBba11eC/test/*

# Block some subpaths with exceptions: last-matching-rule wins (!)
/ipfs/QmUboz9UsQBDeS6Tug1U8jgoFkgYxyYood9NDyVURAY9pK/blocked*
!/ipfs/QmUboz9UsQBDeS6Tug1U8jgoFkgYxyYood9NDyVURAY9pK/blockednot
!/ipfs/QmUboz9UsQBDeS6Tug1U8jgoFkgYxyYood9NDyVURAY9pK/blocked/not
!/ipfs/QmUboz9UsQBDeS6Tug1U8jgoFkgYxyYood9NDyVURAY9pK/blocked/exceptions*

# Block IPNS domain name
/ipns/domain.example

# Block IPNS domain name and path
/ipns/domain2.example/path

# Block IPNS key - blocks wrapped multihash.
/ipns/k51qzi5uqu5dhmzyv3zac033i7rl9hkgczxyl81lwoukda2htteop7d3x0y1mf

# Block all mime types with exceptions
/mime/image/*

> !/mime/image/gif

Reviewer: Is this something that a denylist format needs to support? It feels like it belongs in a local configuration format rather than with something that will get distributed. What's more, media type information is unreliable and rarely available on IPFS as it tends to rely on sniffing.

Reviewer 2: right, I'd break this into two questions that should ideally be decided before pouring cement:
1.) is the system layered enough that local configs and/or processing of lists after merging is a better place for this to economize at higher distro/merging layer, and
2.) is this a required capability or an optional/extension feature (to allow economizing compute at any layer if implementor needs to avoid sniffing anything)

Author: Mime-blocking will likely have to rely on sniffing and likely on stuff that can be interpreted as a file (so unixfs). The concern is warranted. It would also mostly apply to gateways, as it is the main thing turning blocks into files again.

I would be fine striking this off. If someone wants a text-only gateway then it is probably better done with nginx.

Author: Removed mime blocking from spec.

Reviewer 3: @hsanjuan should we remove it from excamples here too?

Reviewer 3: Assuming yes. Removed remaining mentions in 7233af2
We can always add them in future IPIP.

# but not any other paths.
//3f8b9febd851873b3774b937cce126910699ceac56e72e64b866f8e258d09572

# Double hash CID block
# base58btc-sha256-multihash(QmVTF1yEejXd9iMgoRTFDxBv7HAz9kuZcQNBzHrceuK9HR)
# Blocks bafybeidjwik6im54nrpfg7osdvmx7zojl5oaxqel5cmsz46iuelwf5acja
# and QmVTF1yEejXd9iMgoRTFDxBv7HAz9kuZcQNBzHrceuK9HR etc. by multihash
//QmX9dhRcQcKUw3Ws8485T5a9dtjrSCQaUAHnG4iK9i4ceM

# Double hash Path block using blake3 hashing
# base58btc-blake3-multihash(gW7Nhu4HrfDtphEivm3Z9NNE7gpdh5Tga8g6JNZc1S8E47/path)
# Blocks /ipfs/bafyb4ieqht3b2rssdmc7sjv2cy2gfdilxkfh7623nvndziyqnawkmo266a/path
# /ipfs/bafyb4ieqht3b2rssdmc7sjv2cy2gfdilxkfh7623nvndziyqnawkmo266a/path
# /ipfs/f01701e20903cf61d46521b05f926ba1634628d0bba8a7ffb5b6d5a3ca310682ca63b5ef0/path etc...
# But not /path2
//QmbK7LDv5NNBvYQzNfm2eED17SNLt1yNMapcUhSuNLgkqz
```

#### High level list format

A denylist is made of an optional header and a list of blockitems separated by newlines. Comment lines start with `#`. Empty lines are allowed.

```
<header>
---
<block_item1> [hint_list]

# comment
<block_item2> [hint_list]
...
```

#### Header

The list header is a YAML block:

- Must be valid YAML

> - Fully optional

Reviewer: If it's optional but has mandatory syntax, what is the expected behaviour if the syntax is incorrect? Does the implementation ignore the entire list or just the header?

Author: It has known fields that you can expect to find on a header. Missing fields should just be ignored I think.

Author: I added a note saying that failure to parse the header as YAML when it is present (`---` is found) should abort and signal error. A header too-large gets processed as rules, and will error depending if the implementation errors on bad rules or ignored if the implementation ignores bad rules.

> - 1KB maximum size

Reviewer: This seems very short. A description alone could run over. What is the expected processing behaviour in case the size is exceeded?

Author: We can increase the max size to whatever is sensible.

If the denylist does not comply with the spec, then you can consider it invalid and stop processing it as a general rule.

Reviewer: The reason I'm asking all these annoying questions is because I think that this is a spec for which interoperability is vital. If it were just a convenient way of sharing some configuration or something like an adblock list where the worst-case failure is that you see an ad you don't want to see, then the details of implementation behaviour wouldn't matter much. But the failure impact here is that people could end up going to jail, being fined, or having their business blocked, put on the SafeBrowsing list, etc.. As an operator, if I subscribe to a CSAM or DMCA or phishing blocklist, I need to _know_ that it works.

With 10^2 implementations running over 10^6-10^7 nodes (and more!), knowing that it works means we have to be painfully finicky about processing rules. If your implementation reads 1000 bytes before giving up looking for `---` and mine reads 1024, we might get different lists. Or mine assumes UTF-8 (there's no specification of encoding for the format) but you're Japanese and so used UTF-16, we might get wildly different results, or might be counting characters and bytes differently, etc.

I think you specified 1kb here because that's how far you scan for `---`? I think that that's likely to cause problems over time (more metadata gets added leading to needing more space, but can it interoperate with deployed code, version indicators work poorly in distributed environments, etc.). Might an approach like Markdown frontmatter (having a `---` both before and after) not work better?

More generally, would you be okay if I PRed your PR to make it tighter from an interop perspective? It'll be annoying, and it'll make it longer and less pleasant to read, but I'm afraid that we need it for this context.

Author: > I need to _know_ that it works.

The spec cannot ensure this. The implementation may end up failing even if spec is 100% tight on details.

I agree it is good to be more accurate though. What is a reasonable header size?

UTF-8 encoding should also be specified.

> Might an approach like Markdown frontmatter (having a `---` both before and after) not work better?

I don't quite see the point as it doesn't make anything easier with regards to parsing. If `---` is found at the beginning, then you can assume that here is a header but then have to deal with parsing that header. There might still not be a header at all if the denylist is malformed, so you will still have to handle that. You still need to scan until the `---` at the end is found or hit the limit etc.

It is simpler imho the way it is.

It is more interesting to discuss about failure modes: should a malformed entry abort the full list? be ignored? what happens if a header is not found within the limits but the first entry is not a parsable blocking rule (i.e. could signal that the header is too long).

For the the implementation should attempt to block what it can understand and skip over what it cannot understand (logging errors as needed). We should not be blocking or breaking the application due to malformed lists.

Author: I have clarified/added details on how to parse the header. I have increased the max size to 1MiB and specified exactly the number of bytes. I have made explicit that the list is UTF-8 encoded and newlines are `\n` characters.

> - Delimited by a line containing `---` at the end (document separator)

Reviewer: If the header is optional, is the document separator required or are implementations expected to detect that a line beginning with a block item to know that there is no header?

Author: How I do it is to read a max of 1KB looking for `---`. If not there, assume no header and start parsing line by line.

The header is fully optional but if it exists it needs to have the ---.

> Known-fields (they must be lowercase):

Reviewer: Is the intent that these are the only fields that are accepted or are implementations expected to ignore unknown fields? If they ignore unknown fields, is there a way of reserving future standardised additions? If they don't ignore unknown fields, what is the expected processing?

Author: Implementations can expect to find these fields. They may or not be there. There might be other fields too and implementations can do whatever they want with those.

With regards to reserving fields, we could add a rule that any custom fields should start with `x-*` and that implementations should ignore unknown fields otherwise. I don't know if it matters a lot.

Author: I added a note about custom fields starting with `x-`. Custom information can also be added in the hints map. I clarified that header hints are like writing hints for each and every item in the list.

An implementation may return an error when an item is blocked, and ideally, the rule that caused the blocking, and that includes the hints for that rule, both the specific and the header hints.

> - `version`: the denylist format version. Defaults to 1 when not specified.

Reviewer: There doesn't seem to be specific processing happening based on this? Are implementations expected to reject versions they aren't aware of?

Author: Yes, as it would mean they don't know how to support them, I guess. i.e. `version: 2` is not something that is specified, yet.

Author: Added a note that implementations should not parse unsupported versions.

- `name`
- `description`
- `author`
- `hints`: a map of *hints*. See section below for known hints

#### Hints

A *hint* is a key-value duple associated to the denylist as a whole (part of the header), or to a specific \<block_item\>.

Header hints can be used to set denylist-wide options or information that
implementations can choose to interpret or not.

#### List body

A denylist is made of lines which are made by a *block items* followed by zero or more space-separated hints.

> Lines should not be longer than 2MiB including the "\n" delimiter.

Reviewer: Is the expectation that the implementation will bail on longer lists? Or ignore after a cut-off?

Author: It's a limit per line. There is an issue here: you can always have valid IPLD paths that are longer than whatever the limit it, so those would be unblockable. I'm not sure what to make of that. On the other side, having a 50GB line might be a way to exhaust memory. 2MiB seemed like a compromise, but could be higher and be fine too I guess.

Reviewer 2: Wait, does this mean that the *block items*  (referents of the various kinds of entries) need to be under 2MiB, or that lines with 1.9MiB of annotations are valid?

Author: The line of text should be less than 2MiB. How that is divided between block_item and annotations is not specified. IMHO, it is a bit easier for a parser implementation to enforce limits on the basic unit of a line, rather than having separate limits for the block_items or allocations, even if that would perhaps be more "correct".

#### Block item

A block item represents a rule to enable content-blocking:

- `PATH` elements are expected to be %-encoded, per [RFC 3986, section 2.1](https://developer.mozilla.org/en-US/docs/Glossary/percent-encoding).
- `CID` elements represent a CID (either V0 or V1).

> - `CIDv0` are for us equivalent to baseb58btc-encoded sha256 multihashes although they are not the same thing (a CIDV0 carries implicit codec (dag-pb) and multibase information (b58btc). When we say a b58-encoded multihash needs to be extracted from the CID, this usually is a no-op in case of CIDv0s.

Reviewer: Is there specific processing that implementations are expected to carry out for items they don't recognise? Skipping is more resilient to format evolution in a distributed system but I would worry that operators might think they're properly protected by tracking a given list with legal implications when in fact their deployment is skipping over some items it doesn't get.

Author: Implementation-specific? I'm not sure the specification should say how implementations react to things outside the spec. You could kill everything, or you could log an error message, or you could log a warn message. Perhaps we can recommend logging an error.

Author: Added a note

```
$ printf "QmecDgNqCRirkc3Cjz9eoRBNwXGckJ9WvTdmY16HP88768/my/path" | ipfs add --raw-leaves --only-hash --quiet | ipfs cid format -f '%M' -b base58btc
QmSju6XPmYLG611rmK7rEeCMFVuL6EHpqyvmEU6oGx3GR8
```

The rule `//QmSju6XPmYLG611rmK7rEeCMFVuL6EHpqyvmEU6oGx3GR8` will block `/ipfs/bafybeihrw75yfhdx5qsqgesdnxejtjybscwuclpusvxkuttep6h7pkgmze/my/path`, with `QmSju6XPmYLG611rmK7rEeCMFVuL6EHpqyvmEU6oGx3GR8` being the base58-encoded multihash contained in `bafybeihrw75yfhdx5qsqgesdnxejtjybscwuclpusvxkuttep6h7pkgmze`.

We can convert any CID to its multihash with:

```
$ ipfs cid format -f '%M' -b base58btc bafybeihrw75yfhdx5qsqgesdnxejtjybscwuclpusvxkuttep6h7pkgmze
QmecDgNqCRirkc3Cjz9eoRBNwXGckJ9WvTdmY16HP88768
```

Blocking layer recommendation: NameSystem + PathResolver + BlockService.

In order to check for a matching rule, the PathResolver should:

- IPFS path: convert the CID to v1base32 and hash `CIDV1BASE32/PATH` with the
  hashing functions used in the denylist. Match against declared double-hashes.
- IPFS path: convert the CID to CIDv0 and hash `CIDV0/PATH` without trailing `/` with the hashing functions used in the denylist. Match against declared double-hashes.

> - IPNS path:

Author: No, I think it is just necessary to delete that line. The PathResolver never receives an unresolved IPNS path. When I wrote this I didn't know and must have left it like this. I will fix it.

  `$XDG_CONFIG_HOME/ipfs/denylists/` for denylist files.
- Denylist files are processed in alphabetical order so that rules from later
  denylists override rules from earlier denylists on conflict.

### Denylist format

#### Summary

The following example showcases the features and syntax of a compact denylist:

```
version: 1
name: IPFSCorp blocking list
description: A collection of bad things we have found in the universe
author: abuse-ipfscorp@example.com
hints:
  hint: value
  hint2: value2
---
# Blocking by CID - blocks wrapped multihash.
# Does not block subpaths.
/ipfs/bafybeihvvulpp4evxj7x7armbqcyg6uezzuig6jp3lktpbovlqfkuqeuoq

# Block all subpaths
/ipfs/QmdWFA9FL52hx3j9EJZPQP1ZUH8Ygi5tLCX2cRDs6knSf8/*

# Block some subpaths (equivalent rules)
/ipfs/Qmah2YDTfrox4watLCr3YgKyBwvjq8FJZEFdWY6WtJ3Xt2/test*
/ipfs/QmTuvSQbEDR3sarFAN9kAeXBpiBCyYYNxdxciazBba11eC/test/*

# Block some subpaths with exceptions
/ipfs/QmUboz9UsQBDeS6Tug1U8jgoFkgYxyYood9NDyVURAY9pK/blocked*
+/ipfs/QmUboz9UsQBDeS6Tug1U8jgoFkgYxyYood9NDyVURAY9pK/blockednot
+/ipfs/QmUboz9UsQBDeS6Tug1U8jgoFkgYxyYood9NDyVURAY9pK/blocked/not
+/ipfs/QmUboz9UsQBDeS6Tug1U8jgoFkgYxyYood9NDyVURAY9pK/blocked/exceptions*

> 

Reviewer 2: ```suggestion

# Last-wins resolution of conflicts: 

/ipfs/QmUboz9UsQBDeS6Tug1U8jgoFkgYxyYood9NDyVURAY9pK/

!/ipfs/QmUboz9UsQBDeS6Tug1U8jgoFkgYxyYood9NDyVURAY9pK/allowedsubset

/ipfs/QmUboz9UsQBDeS6Tug1U8jgoFkgYxyYood9NDyVURAY9pK/allowedsubset/disallowedsubset

```
(note, I used `!` instead of `+` as negation prefix!)

Author: I'm not sure this is correct, the last line should probably be `/ipfs/QmUboz9UsQBDeS6Tug1U8jgoFkgYxyYood9NDyVURAY9pK/*` as otherwise it won't match `/ipfs/QmUboz9UsQBDeS6Tug1U8jgoFkgYxyYood9NDyVURAY9pK/allowedsubpath` and the path will be allowed (if I'm not mistaken).

I added:

```
# Block some subpaths with exceptions: last-matching-rule wins (!)
```

I think it is illustrative for a showcase of the syntax.

Lines should not be longer than 2MiB including the "\n" delimiter.

#### Block item

A block item represents a rule to enable content-blocking:

- `PATH` elements are expected to be %-encoded, per [RFC 3986, section 2.1](https://developer.mozilla.org/en-US/docs/Glossary/percent-encoding).
- `CID` elements represent a CID (either V0 or V1).
- `CIDv0` are for us equivalent to baseb58btc-encoded sha256 multihashes although they are not the same thing (a CIDV0 carries implicit codec (dag-pb) and multibase information (b58btc). When we say a b58-encoded multihash needs to be extracted from the CID, this usually is a no-op in case of CIDv0s.

##### `/ipfs/CID`

CID-rule: Blocks a specific multihash. If the CID is a V1, it blocks the
multihash contained in it (CIDv0s are multihashes already).

When users want to block by multihash directly, they must base58btc-encoded
multihashes. This rule does not block subpaths that start at this CID, only
the CID itself.

Blocking layer recommendation: BlockService.

##### `/ipfs/CID/PATH`

IPFS-Path-Rule: Blocks the exact ipfs path that is referenced from the

> multihash embedded the CID before attempting to resolve it. It does not block

Reviewer 4: ```suggestion
multihash embedded in the CID before attempting to resolve it. It does not block
```

- `/my/path*` and `/my/path/*`: blocks any paths that contain the prefix `/my/path`.

Blocking layer recommendation: PathResolver.

##### `//DOUBLE_HASH`

Doublehash-Rule: Blocks using double-hashed item, which can be:

- The sha256-hex-encoded hash of `CIDV1_BASE32/PATH`: this is the legacy
  badbits block anchor format. It can only block by CID and not by
  multihash. When no path present, the trailing slash must be kept
  (`CIDV1_BASE32/`).
- A b58-encoded multihash (a.k.a CIDV0), corresponding to the Sum() of:
  - An IPNS-Path:
    - `/ipns/IPNS` when the IPNS name is NOT a CID.
    - The b58-encoded-multihash extracted from an IPNS key when the IPNS key
      is a CID.
  - An IPFS-Path: `b58-encoded-multihash/P/A/T/H` where the multihash is
    extracted from the CID in `/ipfs/CID/P/A/T/H` (The multihash and the CID
    are the same in the case of CIDV0). The `/P/A/T/H` component is optional
    and should not have a trailing `/`.

The latter form allows blocking by double-hash using any hashing function of
choice. Implementations will have to hash requests using all the hashing functions

> used in the denylist, so we recommend sticking to one.

Reviewer 4: Do we need this yet? Can we specify use of a multihash for self-describing future-proof goodness but require that only sha256 is valid for now to simplify things? For context: I'm looking at how i would add support for this spec to e-ipfs and w3s.link.

I acknowledge that we already have to accept providers can only do best effort enforcement of these lists given we already have to deal with path based rules which can't enforced at the BlockService level e.g in e-ipfs which is bitswap only... so i could just ignore entries that are hashed with a fn i don't support yet, but i'd prefer to be able to say "we support all block items that can be applied by the blockservice".

Author: If your denylist only includes one type of hashing function, then you only hash using that one to check the double-hashes. So it is up to who makes the denylist... I would totally discourage using multiple hashing functions for items in the same denylist.

But other than that, since they are multihashes there's no reason to support anything. In a way I'm deferring the discussion of "what is the best hashing function for this" until later, because sha256 is probably not it. Mostly you just want something very very fast, and doesn't necessarily need to be a cryptographic one as it just meant to obfuscate.

Reviewer 2: as above, the tradeoffs are expressiveness of list-sharing for compute-obligation for conforming processors of the lists.  here too deciding layering before committing to a degree of expressiveness (or the core/extension, required/optional corrolary) might be preferable to deciding this in advance of commitments to implement and conform at scale.

Author: I am fine prominently recommending a hashing function. Which one should it be though?

What is:

* A cryptographic hash function (we need this in the end as pre-image and collision protection are important).
* Well supported in all languages/platforms
* Very fast

Is it blake3? sha256? Are we ok with md5 (sure there are collisions but need to be carefully crafted). ? I'll ask around.

Reviewer 4: Would you consider defining it as _only_ sha2-256 for the first version of the spec? That restriction could be loosened later if folks need it.

Author: I would like more input. I think sha2-256 may impose larger perf impact than necessary.

Reviewer 3: Here is an idea/compromise: we explicitly define the hash functions that should be used in the YAML header under `double-hash` key. If the key is not present, we assume SHA256 and clients only use that (backward-compatible). 

This way, we future-proof the format to switch to better functions as the time goes by, but we don't force everyone to waste resources on hashing with every function if the list only uses one.

Example: if we want to limit users to Blake3 `double-hash: 0x1e` or  allow SHA256 and Blake3: `double-hash: 0x1e, 0x12`

Author: Clients don't need to hash with every function. They only hash with the functions that the list uses, and they know which ones are those because of the multihash-prefix, so as you parse the list you know which ones to double-hash with (if any).

The future-proofness is ensured by using multihashes already.

The question is whether we require those multihashes to be sha256 only or something else.

Reviewer 3: For pragmatic reasons, our options are probably limited to either sha2-256 or blake3 (ipfs impls. will have both).
Between sha2 and blake3, the latter should make a faster default (faster than md5 too). Maybe benchmark?

ps. The purpose behind the explicit list of hashes in the YAML header was to protect the node operator from silently exploding the list of hashes to check when the intention is to only use one (or two) functions.  But maybe it is not a real DoS risk, these lists will come from somehow trusted sources.
If we decide that we only use a single hash function per list, it still feels like something that should be explicitly defined in YAML header, and not hard-coded in the spec (future-proofing without having to change the specs).

Reviewer 4: yes, i  strongly agree that the header yaml should define the hash fn, and i am asking that the spec, in it's initial wording, require only sha-256 be used.

we can even add a note in the spec that folks should raise a PR here if that is causing them real perf problems.

adding support for more hash functions would have to take into account that there is no baseline requirement of what hash functions an ipfs impl must support. web3-storage does not verify blake3 today. surprising but true https://github.com/web3-storage/car-block-validator/blob/cb121b780c86724c0266f797b5a4badffad77140/src/index.js#L1-L17

Author: You don't need to list hash functions in the header. As you parse the list, you see the multihash function used in the multihashes that you encounter and that is the list of hashing functions you have double-hash with. Putting the list at the top is redundant.

I think the issue here is a matter of compatibility, and sha256 being the best thing that is nicely and universally supported everywhere. I'm ok with the spec requiring sha256 for double-hashing, otherwise leave things open as they are. If people use other hashes (which remains possible), they would be breaking the spec, but that's up to them.

Reviewer 2: I actually disagree that listing which hash functions are used in a list is necessary and always _redundant_, because I am focused on a use-case where more lists from more unfamiliar authorities get added and subtracted more often, making the list itself worth annotating.  If managing and federating many denylists is being optimized for, all kinds of (optional!) frontmatter fields could be quite helpful in deciding _which ones_ to federate/follow before parsing them, particularly for software that builds on CIDs selectively (which might need to know in advance if a denylist includes entries it cannot, or does not need to, parse).

In any case, I would argue that an optional frontmatter field for declaring which hash functions were/are used to populate a list is good, actually.  I don't have strong opinions, but some basic questions, in case enough people agree that a section is worth adding to the IPIP itself:
1. should the values be `name` entries from the [`multicodec` table](https://github.com/multiformats/multicodec/blob/master/table.csv), or the actual tags?
2. what should a consumer do/think/feel if an entry uses a tag not announced up top?
3. should list hosters ADD entries to the frontmatter declaration over time if they add functions over time, or create a new list at a new URL when these "env var"s change? Considering many subscribers have to unload and reload the list for any changes except EOF appends, this could complicate the spec a lot.

Author: 1. name for sanity
2. With or without hash-function annotations at the top, you're in the same situation if you find a hash that you don't support while parsing. Annotation at the top would only help to reject a list altogether before parsing it. But why would you do that and not try to work with the entries that you can support in any case? Anyways, you can use any optional hints you want. We can standardize an optional hint name for this as long as it's optional.
3. I think you're finding the subtle issues. "whoops now my list includes also blake3 double-hashes" is bad to communicate in any way. "List A is done and you can follow up again on list B" is also not considered. It would be easier if we specify what the recommended behavior is when finding an unknown hashing function.

Also, we are talking a lot about hashing functions pertaining to the double-hash entries that are enconded as multihashes. You could also have normal `/ipfs/<cid>` entries, where the CID itself is not something that your underlying IPFS implementation understands: maybe you don't know the ipld codec, maybe the hash-function is unknown. In this cases, it would not be the "compact denylist parser component/blocker" the one that needs to be aware, but the underlying "ipfs" implementation. etc. but some implications might be similar.

So my practical approach is, 1) everything goes 2) try to parse what you can 3) strong recommendation or obligation to only use sha256 for double-hashing, everything else is non-spec-compliant (for now).

- If NAME is a CID, extract the multihash, encoded with baseb58btc and hash it with the hashing functions used in the denylist. Match against declared double-hashes.

The BlockService should:

- Convert the CID to `CIDV1BASE32/` (keeping the CID codec and adding a slash at the end) and hash it with the hashing functions used in the denylist. Match against declared double-hashes.

- Convert the CID to b58-encoded-multihash (that is CIDv0) and hash the CID string.

##### `/mime/MIMETYPE` `/mime/*`

Blocks content detected to be of the given type. `/mime/*` blocks all the mimetypes and is meant to work with allow rules (all mimetypes blocked except specific ones).

Blocking layer recommendation: Unixfs

Our recommendation is that /mime/ rules automatically set IPFS clients into a
"unixfs only" mode where only unixfs (+raw blocks) are allowed at the
BlockService layer, and content type is checked at the Unixfs layer, as the
blocks get assembled into an actual files. That should cover gateway usage.

#### Allow (or negated) rules

Block items can be prepended by `+`, that items matching the rule are to be allowed

> rather than blocked.

Reviewer 4: Maybe `!` for negation, like a .gitignore

https://git-scm.com/docs/gitignore#_pattern_format

Reviewer 2: In all the examples, negations follow supersets of the negated content without a blank line in between.  this is probably an unintended coincidence, but given that `\n-` and `\n+` are used in some syntaxes like markdown to indent list items or derivative/dependent items, this choice of negation syntax *could* confuse some readers who interpret/assume a rule that you can't negate something unless its a prior entry or subset thereof.

Author: I'm fine changing the negation operator to `!`, (or supporting them all), but then it should be changed across the doc and not in one place.

Author: I have switched to "!"  now.

```
/ipfs/QmecDgNqCRirkc3Cjz9eoRBNwXGckJ9WvTdmY16HP88768/photo*
+/ipfs/QmecDgNqCRirkc3Cjz9eoRBNwXGckJ9WvTdmY16HP88768/photo123.jpg
/mime/*
+/mime/text/plain
+/ipns/my.domain
/ipns/my.domain
```

In this example, `/ipns/my.domain` stays blocked because the deny rule happens
after the allow one.

#### Hint list

A hint list is an optional space-separated list of hints associated with specific block items in the form:

```
<block_item> hintA:v1 hintB:v2 hintC:v3
```

Block items and hints are separated by one or more consecutive instances of
the "space" character.

### Test fixtures

Denylist parsing and correct behaviour can be tested using the
[test.deny](https://github.com/ipfs-shipyard/nopfs/blob/master/tester/test.deny)
denylist, which provides example rules and describes the expected behaviour in
detail.

In particular, a [Blocker implementation validator](https://github.com/ipfs-shipyard/nopfs/tree/master/tester) is provided in Go, and can be adapted to other languages if needed.

### Security

This proposal takes into account security:

- Denylist headers and line-length limits are well specified to avoid malformed lists to cause things like large memory usage while parsing.

> - Supported type of blocks have been though out to avoid amplified consumption of resources or side effects (i.e. downloading of additional dag-blocks) during the implementation.

Reviewer 2: ```suggestion
- Supported type of blocks have been thought out to avoid amplified consumption of resources or side effects (i.e. downloading of additional dag-blocks) during the implementation.
```

as the NameSystem only handles IPNS names (without paths), and the
path.Resolver only handles already-resolved Paths.

##### `/ipns/NAME/*` `/ipns/NAME/PATH*`

IPNS-Path-Prefix-Rule: Same as with the IPFS-Path-Prefix-Rule.

Blocking layer recommendation:  There is no good place to implement this rule
as the NameSystem only handles IPNS names (without paths), and the
path.Resolver only handles already-resolved Paths.

##### `/PATH` `/PATH/*` `/PATH*`

Subpath-Rule: Block solely by looking at the subpath component of an IPFS path. Examples:

- `/my/path`: blocks any item that tries to resolve `/my/path`, regardless of the CID used.
- `/my/path*` and `/my/path/*`: blocks any paths that contain the prefix `/my/path`.

Blocking layer recommendation: PathResolver.

##### `//DOUBLE_HASH`

Doublehash-Rule: Blocks using double-hashed item, which can be:

- The sha256-hex-encoded hash of `CIDV1_BASE32/PATH`: this is the legacy
  badbits block anchor format. It can only block by CID and not by
  multihash. When no path present, the trailing slash must be kept
  (`CIDV1_BASE32/`).
- A b58-encoded multihash (a.k.a CIDV0), corresponding to the Sum() of:
  - An IPNS-Path:

>     - `/ipns/IPNS` when the IPNS name is NOT a CID.

Reviewer 3: ```suggestion
    - `/ipns/name` when the IPNS `name` is NOT a CID.
```

Author: Fixing this by hand.

>     - The b58-encoded-multihash extracted from an IPNS key when the IPNS key
>       is a CID.

Reviewer 3: ```suggestion
    - The b58-encoded-multihash extracted from an IPNS name when the IPNS name
      is a CIDv1 with `libp2p-key` codec.
```

Author: I don't understand the CIDv1 addition. IPNS names can be CIDv0, in which case there is nothing to do, and CIDv1s? Or should I add CIDv0 as a different line?

I didn't care to research if you can have IPNS names using non b58 encoding for CIDv0 or non libp2p-key for CIDv1...

Author: In practice, you try to parse NAME as a CID and if that flies you get the b58-multihash out of it and double hash that.

If not, you double hash `/ipns/name`.

Reviewer 3: This is just a precaution  – better to make spec very explicit and avoid the problems around the fact that PeerID != CIDv0 (CIDv0 only starts with `Q` while PeerID can also start with `1`, see https://cid.ipfs.tech/#12D3KooWHj48PbuAPUf9LeovBJZEXn2A4rseW5z9wbTBwvM82nAq and https://github.com/libp2p/specs/blob/master/peer-ids/peer-ids.md#decoding)

```

The rule `//QmSju6XPmYLG611rmK7rEeCMFVuL6EHpqyvmEU6oGx3GR8` will block
`/ipfs/bafybeihrw75yfhdx5qsqgesdnxejtjybscwuclpusvxkuttep6h7pkgmze/my/path`,
with `QmSju6XPmYLG611rmK7rEeCMFVuL6EHpqyvmEU6oGx3GR8` being the base58-encoded
multihash contained in
`bafybeihrw75yfhdx5qsqgesdnxejtjybscwuclpusvxkuttep6h7pkgmze`.

We can convert any CID to its multihash with:

```
$ ipfs cid format -f '%M' -b base58btc bafybeihrw75yfhdx5qsqgesdnxejtjybscwuclpusvxkuttep6h7pkgmze
QmecDgNqCRirkc3Cjz9eoRBNwXGckJ9WvTdmY16HP88768
```

Blocking layer recommendation: NameSystem + PathResolver + BlockService.

In order to check for a matching rule, the PathResolver should:

- IPFS path: convert the CID to v1base32 and hash `CIDV1BASE32/PATH` with the
  hashing functions used in the denylist. Match against declared double-hashes.
- IPFS path: convert the CID to CIDv0 and hash `CIDV0/PATH` without trailing `/` with the hashing functions used in the denylist. Match against declared double-hashes.
- IPNS path:

The NameSystem should:

- If NAME is a domain name: Hash `/ipns/NAME` with the hashing functions used in the denylist. Match against declared double-hashes.

> - If NAME is a CID, extract the multihash, encoded with baseb58btc and hash it with the hashing functions used in the denylist. Match against declared double-hashes.

Reviewer 3: ```suggestion
- If NAME is a CID, extract the multihash, encode it with baseb58btc and hash it with the hashing functions used in the denylist. Match against declared double-hashes.
```

in future versions.

## Specification

### Denylist file extension, locations and order

While not pertaining to the denylist format itself, we introduce the following conventions about denylist files when they are stored in the local filesystem:

- Denylist files are named with the extension `.deny`.
- Implementations should look in `/etc/ipfs/denylists/` and
  `$XDG_CONFIG_HOME/ipfs/denylists/` for denylist files.
- Denylist files are processed in alphabetical order so that rules from later
  denylists override rules from earlier denylists on conflict.

### Denylist format

#### Summary

The following example showcases the features and syntax of a compact denylist:

```
version: 1
name: IPFSCorp blocking list
description: A collection of bad things we have found in the universe
author: abuse-ipfscorp@example.com
hints:
  hint: value
  hint2: value2
---
# Blocking by CID - blocks wrapped multihash.
# Does not block subpaths.
/ipfs/bafybeihvvulpp4evxj7x7armbqcyg6uezzuig6jp3lktpbovlqfkuqeuoq

# Block all subpaths
/ipfs/QmdWFA9FL52hx3j9EJZPQP1ZUH8Ygi5tLCX2cRDs6knSf8/*

# Block some subpaths (equivalent rules)
/ipfs/Qmah2YDTfrox4watLCr3YgKyBwvjq8FJZEFdWY6WtJ3Xt2/test*

> /ipfs/QmTuvSQbEDR3sarFAN9kAeXBpiBCyYYNxdxciazBba11eC/test/*

Author: Say you are running a gateway only, perhaps the only place you want to block is at the namesys and resolver levels.

So if you put `/ipfs/cid`  in the denylist, it would still be fine as long as you  are not blocking in the BlockService-level.

Our gateways/kubo need more comprehensive blocking (as, I don't want to "fetch" a blocked CID in the first place), but I can imagine applications which are ok only blocking by path and not by block, and allow to fetch blocks to resolve paths, even if paths to those specific blocks are blocked.

Also note that if "/ipfs/cid" is blocked, we see this issue with subpaths. But if we are talking of "/ipfs/cid/path" and "/ipfs/cid/path/other", the latter might still work, depending on how resolution is implemented. So `/*`  has a point there.

> /ipfs/QmTuvSQbEDR3sarFAN9kAeXBpiBCyYYNxdxciazBba11eC/test/*

Reviewer 3: Thanks, this makes sense, I think the problem here is that this is useless example of the path blocking:

```
# Block all subpaths
/ipfs/QmdWFA9FL52hx3j9EJZPQP1ZUH8Ygi5tLCX2cRDs6knSf8/*
```

We have better example below, with actual subpath:

```
# Block some subpaths (equivalent rules)
/ipfs/Qmah2YDTfrox4watLCr3YgKyBwvjq8FJZEFdWY6WtJ3Xt2/test*
```

I think we should just remove the `/ipfs/cid/*` example as it is only mudding the water.
`/ipfs/cid/subdpath/*` is way better.

> /ipfs/QmTuvSQbEDR3sarFAN9kAeXBpiBCyYYNxdxciazBba11eC/test/*

Author: I have removed the example and added NOTEs explaning and warning about the behaviour here.

!/ipfs/QmUboz9UsQBDeS6Tug1U8jgoFkgYxyYood9NDyVURAY9pK/blockednot
!/ipfs/QmUboz9UsQBDeS6Tug1U8jgoFkgYxyYood9NDyVURAY9pK/blocked/not
!/ipfs/QmUboz9UsQBDeS6Tug1U8jgoFkgYxyYood9NDyVURAY9pK/blocked/exceptions*

# Block IPNS domain name
/ipns/domain.example

# Block IPNS domain name and path
/ipns/domain2.example/path

# Block IPNS key - blocks wrapped multihash.
/ipns/k51qzi5uqu5dhmzyv3zac033i7rl9hkgczxyl81lwoukda2htteop7d3x0y1mf

# Block all mime types with exceptions
/mime/image/*
!/mime/image/gif

# Legacy CID double-hash block
# sha256(bafybeiefwqslmf6zyyrxodaxx4vwqircuxpza5ri45ws3y5a62ypxti42e/)
# blocks only this CID
//d9d295bde21f422d471a90f2a37ec53049fdf3e5fa3ee2e8f20e10003da429e7

# Legacy Path double-hash block
# Blocks bafybeiefwqslmf6zyyrxodaxx4vwqircuxpza5ri45ws3y5a62ypxti42e/path
# but not any other paths.
//3f8b9febd851873b3774b937cce126910699ceac56e72e64b866f8e258d09572

# Double hash CID block
# base58btc-sha256-multihash(QmVTF1yEejXd9iMgoRTFDxBv7HAz9kuZcQNBzHrceuK9HR)
# Blocks bafybeidjwik6im54nrpfg7osdvmx7zojl5oaxqel5cmsz46iuelwf5acja
# and QmVTF1yEejXd9iMgoRTFDxBv7HAz9kuZcQNBzHrceuK9HR etc. by multihash
//QmX9dhRcQcKUw3Ws8485T5a9dtjrSCQaUAHnG4iK9i4ceM

> # Double hash Path block using blake3 hashing
> # base58btc-blake3-multihash(gW7Nhu4HrfDtphEivm3Z9NNE7gpdh5Tga8g6JNZc1S8E47/path)
> # Blocks /ipfs/bafyb4ieqht3b2rssdmc7sjv2cy2gfdilxkfh7623nvndziyqnawkmo266a/path
> # /ipfs/bafyb4ieqht3b2rssdmc7sjv2cy2gfdilxkfh7623nvndziyqnawkmo266a/path
> # /ipfs/f01701e20903cf61d46521b05f926ba1634628d0bba8a7ffb5b6d5a3ca310682ca63b5ef0/path etc...
> # But not /path2
> //QmbK7LDv5NNBvYQzNfm2eED17SNLt1yNMapcUhSuNLgkqz

Reviewer 3: @hsanjuan `QmbK7LDv5NNBvYQzNfm2eED17SNLt1yNMapcUhSuNLgkqz` uses sha256, not blake3, blake3 is only used in the root CID of th epath that is being blocked (`bafyb4ieqht3b2rssdmc7sjv2cy2gfdilxkfh7623nvndziyqnawkmo266a`) is this intended?

Would it make sense to switch this to multihash which is also blake3, to demonstrate one can use something other than sha256?

below generated with `printf "gW7Nhu4HrfDtphEivm3Z9NNE7gpdh5Tga8g6JNZc1S8E47/path" | ipfs add --raw-leaves --only-hash --hash blake3 --quiet | ipfs cid format -f '%M' -b base58btc`: 

```suggestion
# Double hash Path block using blake3 hashing
# base58btc-blake3-multihash(gW7Nhu4HrfDtphEivm3Z9NNE7gpdh5Tga8g6JNZc1S8E47/path)
# Blocks /ipfs/bafyb4ieqht3b2rssdmc7sjv2cy2gfdilxkfh7623nvndziyqnawkmo266a/path
# /ipfs/bafyb4ieqht3b2rssdmc7sjv2cy2gfdilxkfh7623nvndziyqnawkmo266a/path
# /ipfs/f01701e20903cf61d46521b05f926ba1634628d0bba8a7ffb5b6d5a3ca310682ca63b5ef0/path etc...
# But not /path2
//gW813G35CnLsy7gRYYHuf63hrz71U1xoLFDVeV7actx6oX
```

Author: ok...  The spec equates base58btc with CIDv0 but blake3 multihash should work too.

This is wrong in the implementation. It assumes the double-hash is a CIDv0.
